RFC for Boxes (#1)
SamFerracin/RFCs: 0022-circuit-block.md

# Introducing `Block`s in qiskit

| **Status**        | **Proposed** |
|:------------------|:---------------------------------------------|
| **RFC #**         | 0022                                         |

> | **Authors**       | [Sam Ferracin](sam.ferracin@ibm.com), [Ian Hincks](ian.hincks@ibm.com), [Chris Wood](cjwood@ibm.com), [Jake Lishman](jake.lishman@ibm.com), [Joshua Skanes Norman](joshua.sn@ibm.com)    |

Reviewer: ```suggestion
| **Authors**       | [Sam Ferracin](sam.ferracin@ibm.com), [Ian Hincks](ian.hincks@ibm.com), [Chris Wood](cjwood@ibm.com), [Jake Lishman](jake.lishman@ibm.com), [Joshua Skanes-Norman](joshua.sn@ibm.com)    |
```

| **Submitted**     | YYYY-MM-DD                                   |

## Summary
Objective: Introduce a new `Block` instruction to enhance the user experience with Qiskit's Runtime primitives, with the goal of providing more flexibility and transparency with respect to twirling and mitigation.

> The concept of a block of gates is fundamental to many quantum computing routines. For example, a block is a natural concept when reasoning about quantum errors because the error profile often depends on factors such as which gates are applied simultaneously. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks *indirectly* using barriers. While this approach has been adequate for tasks like scheduling and transpilation, barriers can be hard to parse when pre-processing circuits in preparation for twirling or mitigation. Suboptimal handling of barriers in these post-processing steps can introduce significant slowdowns in some mitigation experiments (for example, it can unnecessarily increase the time required by the noise-learning steps) and produce outcomes that may seem unintuitive to typical users.

Reviewer: ```suggestion
The concept of a block of gates is fundamental to many quantum computing routines. For example, a block is a natural concept when reasoning about quantum errors because the error profile often depends on factors such as which gates are applied simultaneously. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks *indirectly* using barriers. While this approach has been adequate for tasks like scheduling and transpilation, barriers can be hard to parse when pre-processing circuits in preparation for twirling or mitigation. Suboptimal handling of barriers in these post-processing steps can introduce significant slowdowns in some mitigation experiments (for example, it can unnecessarily increase the time required by the noise learning steps) and produce outcomes that may seem unintuitive to typical users.
```

> To overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new `Block` class. The primary goal of this class is to encapsulate an isolated block of `CircuitInstruction`s, treating it as a single unit for tasks like twirling and mitigation. We believe that `Block`s will offer:

Reviewer 2: ```suggestion
To overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new `Block` instruction class. The primary goal of this class is to encapsulate an isolated block of `CircuitInstruction`s, treating it as a single unit for tasks like twirling and mitigation. We believe that `Block`s will offer:
```

> The overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new `Block` class. The primary goal of this class is to encapsulate an isolated block of `CircuitInstruction`s, treating it as a single unit for tasks like twirling and mitigation. We believe that `Block`s will offer:

Reviewer: ```suggestion
To overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new `Block` class. The primary goal of this class is to encapsulate an isolated block of `CircuitInstruction`s, treating it as a single unit for tasks like twirling and mitigation. We believe that `Block`s will offer:
```

* Full transparency: 
  * Users will be able to learn how their circuits are broken into blocks, twirled, and mitigated *before* running a job
  (today, they can only find this out once the job is done).
* Full flexibility: 
  * Users will be able to draw blocks around every combination of instructions that they want to be grouped together for the purpose of twirling or mitigation (today, all blocks are effectively layers, i.e., depth-one blocks).
* Ability to specify a twirling context:
  * Users will be able to define a twirling group such as "Pauli" and a twirling strategy such as "active-accum" that can potentially be different for each block (today, this can only be done at the circuit level).

> * Ability to assign a noise handle to each block that uniquely specifies its noise model (today, this requires hashing circuits and is inherently unreliable, since circuits are mutable objects).

Reviewer 2: ```suggestion
* Transparency: 
  * Users will be able to learn how their circuits are broken into blocks, twirled, and mitigated *before* running a job
  (today, they can only find this out once the job is done).
* Flexibility: 
  * Users will be able to draw blocks around any combination of instructions that they want to be grouped together for the purpose of twirling or mitigation (today, all blocks are effectively layers, i.e., depth-one blocks).
* Extensibility:
  * A block instance can own a context that can, for example, allow users to specify a twirling group such as "Pauli" and a twirling strategy such as "active-accum" that can potentially be different for each block (today, this can only be done at the circuit level). We can introduce allowed features to the context as necessary, and have definitions for what features the runtime supports.
* Ability to assign a handle to each unique block that enables user to target noise models against handles to be used in mitigation or simulation (today, there is no clean way of doing this, and usually involves an unsafe hash of a quantum circuit object, fragile indexing, or human introspection).
```

## `Block`: Definition and use cases

> A `Block` represents an isolated block of `CircuitInstruction`s that is treated as a single unit in the context of twirling and mitigation. In more detail, when a circuit contains blocks:

Reviewer 2: ```suggestion
A `Block` represents an isolated block of `CircuitInstruction`s that is treated as a single unit in the context of twirling and mitigation. A circuit executor, such as the runtime primitives, should implement all like blocks acting on the same set of qubits in the same way. In more detail, when a circuit contains blocks:
```

> - Twirling targets the blocks and ignores all the circuit instructions that live outside of the blocks.

Reviewer 2: However, the twirling interpreter responsible for execution will need to look at things outside the block and do things like single-qubit gate merging, no?

Author: Yes, I agree with you

- Noise learning targets the blocks and ignores all the circuit instructions that live outside of the blocks.
- Gate mitigation (e.g. PEC and PEA) targets the blocks and ignores all the circuit instructions that live outside of the blocks.

We envision two main workflows that users will follow once `Block`s become available, one for regular users and another one for power users:
1. The workflow for regular users:
    - Users initialize a `QuantumCircuit` without blocks, as they do today.
    - They apply all of the desired transiler passes, for example to map the circuit to an ISA circuit for the backend that they wish to use.
    - They use (a convenience method built around) a new transpiler pass that collects circuit's gates into blocks, with the ability of specifying different collection strategies.
    - They submit their job.
2. The workflow for power users:

>     - Users initialize a `QuantumCircuit` adding blocks manually as they wish.

Reviewer: What does `as they wish` mean specifically? For a regular user, the workflow implies the blocks get added after the circuit is mapped to an ISA circuit. Is the idea that a power user could specify the circuit and blocks at the ISA level or block the circuit before the mapping?

Author: Yes, exactly.
It's not hard to directly write a circuit that complies directly with ISA requirements, you just need to use the right gates and have the correct coupling map in front of you. Power users can do this, or also, apply their own passes to the sub-circuits and then inserting them into the bigger circuit as blocks

print(my_block[0])
>>> "CircuitInstruction(operation=Instruction(name='block', num_qubits=2, num_clbits=0, params=[]), qubits=(Qubit(QuantumRegister(3, 'qreg1'), 1), Qubit(QuantumRegister(3, 'qreg2'), 2)), clbits=())"

print(circuit.draw())
>>> qreg1_0: ────────────────
>>>          ┌───┐┌────────┐
>>> qreg1_1: ┤ S ├┤0       ├─
>>>          └───┘│        │
>>> qreg1_2: ─────┤        ├─
>>>               │        │
>>> qreg2_0: ─────┤  Block ├─
>>>          ┌───┐│        │
>>> qreg2_1: ┤ H ├┤        ├─
>>>          └───┘│        │
>>> qreg2_2: ─────┤1       ├─
>>>               └────────┘
```

## Outstanding questions:
1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the `NoiseLearner` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. How can we design the noise handle so that it is safer than that, but also more flexible (e.g., if users want to mutate the block but still assign the same noise to it during mitigation, they should be allowed to do so)?

> 2. Should `Block` added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?

Reviewer: ```suggestion
2. Should `Block` be added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?
```

> 3. What packages and subpackages need to be modify once `Block` are introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?

Reviewer: ```suggestion
3. What packages and subpackages need to be modified once `Block` is introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
```

> 4. How can we (should we?) facilitate re-use of CircuitInstruction or Block instances? For example, suppose we want the same block in a circuit 100 times:

Reviewer 2: ```suggestion
4. How and when do we communicate the "scope of a block context"? For example, suppose a block context is only sensibly defined for depth-1 blocks of entanglers. Do we let this thing be instantiated with anything and raise during execution, or do we raise immediately? Who is in charge of validation?
5. How can we (should we?) facilitate re-use of CircuitInstruction or Block instances? For example, suppose we want the same block in a circuit 100 times:
```

- Users of the primitives can inspect the blocks before submitting their jobs. This is a big improvement: today, the block generation happens in the server and can be unintuitive.

In addition to supporting the two workflow above, we believe that the following workflow should *not* be supported:

3. The workflow for "lazy" users:
    - Users initialize a `QuantumCircuit` without blocks, as they do today.
    - They apply all of the desired transiler passes, for example to map the circuit to an ISA circuit for the backend that they wish to use.
    - They submit their job.
    - The server creates blocks for the users "behind the scenes."

This workflow essentially takes the same steps as are taken today and therefore presents the same disadvantages discussed above, and also violates the principle "don't do hidden transpilation server-side".

Having described motivation and use cases for circuit blocks, we can now provide details regarding their implementation.

## `Block` as an `Instruction`
In this implementation, `Block` are `Instruction` objects that can be initialized from a circuit.
```python

> class Block(Instruction):

Reviewer 3: [comment 1]

I would like to call this `Box`, since I think it aligns very closely with the OpenQASM 3 concept of a `box`.

    """A container for an isolated block of operations in a larger circuit.
    
    Args:
        circuit: A quantum circuit containing all the operations in the block.
        twirling_strategy: Some specification of a twirling group (such as "pauli")
            and of a twirling mode (such as "active-accum") to apply to this block.
        noise_handle: A mapping object that maps a unique identifier of this block to a noise model.
        label: A label.
    """

    def __init__(self, circuit: QuantumCircuit, twirling_strategy, noise_handle, label: str = ""):
        self._circuit = circuit
        self._twirling_strategy = twirling_strategy
        self._noise_handle = noise_handle

        super().__init__(
            "block",
            self.circuit.num_qubits,
            self.circuit.num_clbits,
            params=self.circuit.parameters,
            label=label,
        )

> ```

Reviewer 3: [comment 2]

I think this should inherit from `ControlFlowOp`, not raw `Instruction`.  If it's raw `Instruction`, we have to re-invent a whole bunch of the special-case handling that was already super painful to teach the core library, except now it'll be slightly different as well, so even more complex.

As a few examples from the top of my head:

- all transpiler passes would immediately need to learn how to skip such an instruction
- ISA circuits checks would need special-case handling for this one instruction, since it doesn't behave the same way as other "container" objects that already exist
- the circuit visualisers would need to learn to handle this specially as well
- we'd have to invent new syntax for the block-construction phase in Qiskit, if we want to go beyond the "create another circuit and `append` it" form
- `assign_parameters` would need to be taught new logic to understand how to recurse into the block

Being `Instruction`s, they can be appended to `QuantumCircuit`s simply by specifying the lists of `qargs` and `clargs`, for example via a dedicated `block` function wrapping the `append` method of `QuantumCircuit`:
```python
class QuantumCircuit:
    ...

    def block(
        self, 
        circuit: QuantumCircuit,
        ...,  # the other inputs required by `Block`s
        qargs: Optional[list[QubitSpecifier]] = None,
        cargs: Sequence[ClbitSpecifier] | None = None,
    ) -> InstructionSet:
        from .block import Block

        # These use a `dict` not a `set` to guarantee a deterministic order to the arguments.
        qargs = tuple(
            {q: None for qarg in qargs for q in self._qbit_argument_conversion(qarg)}
        )
        cargs = tuple(
            {b: None for carg in cargs for b in self._clbit_argument_conversion(carg)}

>         )

Reviewer 3: It's neither here nor there if we swap to using a `ControlFlowOp` (since the builders handle the arguments for you), but internally casting the inputs to a set-like object will mask problems or produce surprising results if the user submits `qargs` that accidentally name the same qubit more than once.  It's not super clear to me why we'd try and remove duplicates automatically here; why would the user _try_ to include duplicates in an ordered list?

        return self.append(Block(circuit, ...), qargs, cargs)

> ```

Reviewer 3: This is a similar-but-not-identical version of what we already have for control-flow operations.

2. Should `Block` be added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?
3. What packages and subpackages need to be modified once `Block` is introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
4. How and when do we communicate the "scope of a block context"? For example, suppose a block context is only sensibly defined for depth-1 blocks of entanglers. Do we let this thing be instantiated with anything and raise during execution, or do we raise immediately? Who is in charge of validation?
5. How can we (should we?) facilitate re-use of CircuitInstruction or Block instances? For example, suppose we want the same block in a circuit 100 times:
```python
for _ in range(100):
    with circuit.block(context=my_context, name="foo") as block:
        block.h(0) 
        block.cx(0, 1)

> ```

Reviewer 3: Some comments on the open questions:

1. making a block hashable within a context where you _expect_ to mutate them sounds like a recipe for horrible bugs at some point in the future.  If all you need is a handle to a block, and you don't want equivalent blocks to map to the same key, then just assign a UUID in the metadata of a block that you look up.  If you do, then define a "key function" that maps blocks to some UUID, and store the resulting UUIDs as the keys in the dict, rather than the blocks themselves.
2. imo, the `Box` I sketched above is suitable for Qiskit core, but not the pec-runtime-specific `Block` with `noise_key` and `twirling_strategy` that this RFC currently describes.
3. If we make this a `ControlFlowOp`, we'll be able to inherit a lot of the hardest work in other packages, but there's still some changes needed.  Off the top of my head:
   * Aer needs to be taught to effectively just unwrap the `block` as part of its control-flow-operation compilation.  This is easy.
   * `qiskit-ibm-runtime` and the IBM backends need to communicate that they support the control-flow op `box` in their `Target`s, just like they do for other control flow
4. These concerns sound very specific to pec-runtime's current implementation of twirling.  Certainly it seems inappropriate to constrain Qiskit core's implementation of `box` based on them.
5. As mentioned in the top comment, I really very much want to separate this idea of `box` from the idea of re-use.  There's really _loads_ of complexity in what it truly means to "re use" a structural element of a circuit (as opposed to just the unitary action of a gate, which is the sort of re-use we handle nowadays).  I think this `box` should be considered entirely separately to the question of re-use, and we should make it clear that `box` is the primitive that solves "grouping with a context" and not the primitive that solves the (somewhat unrelated) problem of "reuse of program structure".

Reviewer 2: 4. Issue 4 can be more generally stated as: "can the annotations of a box have an impact on the allowed contents of a box at construction time?" Offline, a subset of us discussed whether or not this kind of thing would be more appropriate during primitive validation.

5. I'm not sure exactly what you mean by reuse. The Box concept being discussed is reusable in the sense that, like every other kind of instruction, there would be nothing stopping someone from adding the same box instance multiple times to a single circuit. By reuse, do you mean that tools such as the transpiler would be forced to perform the exact same operation on every like block encountered? Like if it scheduled the contents of a box in one way, if it ever encountered the same block again, it would by contract have to schedule the new one in the same way?

# Introducing `CircuitBlock`s in qiskit

| **Status**        | **Proposed** |
|:------------------|:---------------------------------------------|
| **RFC #**         | 0022                                         |
| **Authors**       | [Sam Ferracin](sam.ferracin@ibm.com), [Ian Hincks](ian.hincks@ibm.com), [Chris Wood](cjwood@ibm.com), [Jake Lishman](jake.lishman@ibm.com), [Joshua Skanes Norman](joshua.sn@ibm.com)    |
| **Submitted**     | YYYY-MM-DD                                   |

## Summary

> Objective: Introducing a new ``CircuitBlock`` class to enhance the user experience with Qiskit's built-in primitives, with the goal of providing more flexibility and full transparency with respect to twirling and mitigation.

Reviewer 2: ```suggestion
Objective: Introduce a new ``CircuitBlock`` instruction to enhance the user experience with Qiskit's Runtime primitives, with the goal of providing more flexibility and transparency with respect to twirling and mitigation.
```

Reviewer 2: Separately, I might just call it a `Block`. 

Also, incidentally, this is markdown so you want single backticks.

> The concept of a "block" of gates—meaning a subset of gates isolated from the rest of the circuit—is fundamental to many quantum computing routines. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks *indirectly* using barriers. While this approach has been adequate for tasks like scheduling and transpilation, barriers can be hard to parse when pre-processing circuits in preparation for twirling or mitigation. Suboptimal handling of barriers in these post-processing steps can introduce significant slowdowns in some mitigation experiments and produce outcomes that may seem unintuitive to typical users.

Reviewer 2: I have some words that might somehow be edited into this paragraph:

A block of gates is a collection of circuit instructions that is promised to have a consistent execution when applied to a particular set of qubits. For example, the relative scheduling of gates within a block is consistent between multiple applications of the block. A circuit layer, or cycle, is a special case of a block that is full-width and depth-1.

In particular, a block is a natural concept when reasoning about quantum errors because the error profile often depends on factors such as which gates are applied simultaneously.

Author: I stole the second sentence. Feel free to add the first one too if you find a natural spot for it :)

> The overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new ``CircuitBlock`` class. The primary goal of this class is to encapsulate an isolated block of ``CircuitInstruction``s, treating it as a single unit for tasks like twirling and mitigation. We believe that ``CircuitBlock``s will offer greater transparency in pre-processing routines for twirling and mitigation, increased flexibility in how circuits are twirled or mitigated, and improved runtimes for certain mitigation experiments.

Reviewer 2: ```suggestion
To overcome these issues, we propose to establish the concept of a block of gates through the introduction of a new ``CircuitBlock`` instruction class. The primary goal of this class is to encapsulate an isolated block of ``CircuitInstruction``s, treating it as a single unit for tasks like twirling and mitigation. We believe that ``CircuitBlock``s will offer greater transparency in pre-processing routines for twirling and mitigation, increased flexibility in how circuits are twirled or mitigated, and improved runtimes for certain mitigation experiments.
```

> ## Motivation

Reviewer 2: Reading this motivation section, I am expecting that it will come across as unmotivating for people not intimately familiar with the idiosyncratic issues of block collection in pec-runtime, that is, just about all non-author readers. 

I think we could write a stronger motivation section if we focused on the idea that a user won't know how we collect blocks until they get results, rather than this particular enumeration. I expect the common response to this section will be "duh, 3 is obviously best; do greedy collection"

To better illustrate the issue at hand, consider applying learning-based mitigation (e.g., PEA) to a circuit that contains barriers, e.g.
```
      ░       ░       ░       ░       ░       ░       ░ 
q_0: ─░───■───░───■───░───■───░───■───░───■───░───■───░─
      ░ ┌─┴─┐ ░ ┌─┴─┐ ░ ┌─┴─┐ ░ ┌─┴─┐ ░ ┌─┴─┐ ░ ┌─┴─┐ ░ 
q_1: ─░─┤ X ├─░─┤ X ├─░─┤ X ├─░─┤ X ├─░─┤ X ├─░─┤ X ├─░─
      ░ └───┘ ░ └───┘ ░ └───┘ ░ └───┘ ░ └───┘ ░ └───┘ ░ 
```

> To perform the noise-learning part of the protocol, the mitigation algorithm needs to break this circuit into blocks, and in order to do so, it can choose to handle barriers in three different ways:

Reviewer 2: ```suggestion
To perform the noise-learning part of the protocol, the mitigation algorithm needs to break this circuit into blocks, and in order to do so, it might choose to handle barriers in three different ways:
```

> 1. Including the barriers into the blocks (as is currently done).
> 2. Ignoring the barriers.
> 3. Switching to a new blocks when a barrier is encountered.

Reviewer 2: ```suggestion
1. Include the barriers into the blocks (as is currently done).
2. Ignore the barriers.
3. Switch to a new block when a barrier is encountered.
```

    - They use (a convenience method built around) a new transpiler pass that collects circuit's gates into blocks, with the ability of specifying different collection strategies.
    - They submit their job.
2. The workflow for power users:
    - Users initialize a ``QuantumCircuit`` adding blocks manually as they wish.
    - They submit their job.

Setting up these two workflows presents notable advantages, such as:
- All of the existing transpiler passes can be asked (at least initially) to simply ignore the blocks, since transpilation is meant to happen before the blocks are generated. This can save quite a lot of time that would otherwise be spent adding logic to every single transpiler pass -- and yet, we can choose to do this in the future.
- Users of the primitives can inspect the blocks before submitting their jobs. This is a big improvement: today, the block generation happens in the server and can be unintuitive (see the example in the background section), and users have to wait for the job to be over before they can find out how the circuit was broken into blocks.

> In addition to supportin the two workflow above, we believe that the following workflow should not be supported:

Reviewer 2: ```suggestion
In addition to supportin the two workflow above, we believe that the following workflow should *not* be supported:
```

3. The workflow for "lazy" users:
    - Users initialize a ``QuantumCircuit`` without blocks, as they do today.
    - They apply all of the desired transiler passes, for example to map the circuit to an ISA circuit for the backend that they wish to use.
    - They submit their job.
    - The server creates blocks for the users "behind the scenes."

> This workflow essentially takes the same steps as are taken today and therefore presents the same disadvantages as we have today, thus we believe that is should be banned.

Reviewer 2: ```suggestion
This workflow essentially takes the same steps as are taken today and therefore presents the same disadvantages discussed above, and also violates the principle "don't do hidden transpilation server-side".
```

circuit.h(qreg2[1])
my_block = circuit.block(block, [1, 5])
# equivalent to: my_block = circuit.block(block, [qreg1[1], qreg2[2]])

print(my_block[0])
>>> "CircuitInstruction(operation=Instruction(name='block', num_qubits=2, num_clbits=0, params=[]), qubits=(Qubit(QuantumRegister(3, 'qreg1'), 1), Qubit(QuantumRegister(3, 'qreg2'), 2)), clbits=())"

print(circuit.draw())
>>> qreg1_0: ────────────────
>>>          ┌───┐┌────────┐
>>> qreg1_1: ┤ S ├┤0       ├─
>>>          └───┘│        │
>>> qreg1_2: ─────┤        ├─
>>>               │        │
>>> qreg2_0: ─────┤  Block ├─
>>>          ┌───┐│        │
>>> qreg2_1: ┤ H ├┤        ├─
>>>          └───┘│        │
>>> qreg2_2: ─────┤1       ├─
>>>               └────────┘
```

Similar to `for_loop` and `switch`, blocks can also be appended within a context:
```
with circuit.block(context=my_context, name="foo") as block:
    block.h(0) 
    block.cx(0, 1) 
    block.cx(5, 6) # note that for_loop, switch and other ControlFlowOps have magic methods that
                   # lets these indices, (5,6) here, be wrt to `circuit`, converting them to
                   # appropriate input to Block
```

## Outstanding questions:

> 1. In the existing implementation of blocks that exists in `pec-runtime`, blocks are hashable, so that they can be used as keys in dictionaries (specifically in the dictionary porduced by the ``NoiseLearner`` that maps blocks to noise models). This is clearly unsafe since circuits are mutable. Can we find a way around this? 

Reviewer 2: In particular, how are users going to tie noise models to blocks if we can't have a `dict[CircuitInstruction[Block], NoiseModel]` or equivalent. This is a crucial feature, so we need to have this planned for.

2. Should ``CircuitBlock`` added to qiskit core, alongside the transpiler pass groups gates in blocks workflow 1? If so, should they be written in Rust or Python?

> 3. What packages and subpackages need to be modify once ``CircuitBlocks`` are introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?

Reviewer 2: ```suggestion
3. What packages and subpackages need to be modify once ``CircuitBlocks`` are introduced? For example, should we add logic to qiskit-aer to be able to simulate circuits with blocks, or to Qiskit's optimizer, ..?
4. How can we (should we?) facilitate re-use of CircuitInstruction or Block instances? For example, suppose we want the same block in a circuit 100 times:
```python
for _ in range(100):
    with circuit.block(context=my_context, name="foo") as block:
        block.h(0) 
        block.cx(0, 1) 
```

| **Authors**       | [Sam Ferracin](sam.ferracin@ibm.com), [Ian Hincks](ian.hincks@ibm.com), [Jake Lishman](jake.lishman@ibm.com), [Joshua Skanes-Norman](joshua.sn@ibm.com)    |
| **Submitted**     | YYYY-MM-DD                                   |

## Summary

Introduce a way of grouping instructions in Qiskit SDK that can have downstream data attached, pass through transpilation, and up and down an execution stack.
The case-study use of this is for Qiskit Runtime primitives to provide more flexibility and transparency in Pauli twirling and error mitigation.

Qiskit does not currently have the concept of a "grouping" of instructions, other than as a user-defined custom gate.
However, the ability to group certain instructions to appear as a "block" in a larger context is important for many compilation tasks:

* a block of several gates might have a known unitary action, and can be treated as atomic for commutation purposes, even if a candidate gate doesn't commute with each block element individually.
* noise-learning techniques might be applied to simultaneous logical blocks of operations, rather than individual 2q hardware gates.

> * a user might wish to Pauli twirl around a block of instructions, rather than a single one.

Reviewer 2: ```suggestion
* a user might wish to Pauli twirl around a block of instructions, including idling wires, rather than a single one.
```

* a block of instructions might be assigned some duration as part of a larger circuit, and internally it is scheduled with stretchable durations to implement optimal dynamic decoupling spacing without increasing the complexity of the outer scheduling problem.

Several downstream projects of Qiskit have made ad-hoc blocks by using labelled `Barrier` instructions, but this optimisation barrier has further effects than simple grouping, and the single-sided nature of it makes it very hard to work with as an actual "grouping" construct.

> `Barrier`s also are very hard for users to interpret when inspecting circuits after compilation.

Reviewer 2: superfluous superlative
```suggestion
`Barrier`s also are hard for users to interpret when inspecting circuits after compilation.
```

> The RFC proposes to add a new `Box` instruction to Qiskit, whose semantics will be very similar to the `box` concept from OpenQASM 3.

Reviewer 2: ```suggestion
The RFC proposes to add a new `Box` instruction to Qiskit, whose semantics will be very similar to the `box`
[concept from OpenQASM 3](https://openqasm.com/language/scope.html#block-scope).
```

This `Box` should be able to pass through the Qiskit compiler (under certain limitations), be QPY serialisable, and be able to be transmitted up and down execution stacks.
We will examine how this `Box` might also permit downstream "annotations", outside of the control of the Qiskit SDK compiler, and how the compiler will still be able to reason about such annotations.

The nearest-term use cases for this are:

* have a grouping construction to simplify stretch-based scheduling in Qiskit SDK
* allow advanced users to specify the noise-learning and twirling blocks IBM Primitives error mitigation should use, and the per-block strategy
* allow regular users to see what blocks the IBM Primitives divided their circuit into for noise learning and twirling

## `Box`: a one-off grouping of instructions

As in OpenQASM 3, a `Box` will be:

> * a group of non-overlapping (but potentially nested) circuit instructions,

Reviewer 2: I'm not sure that the OQ3 spec actually does talk about contained instructions being non-overlapping, and I'm not sure why it would for the same reason I wouldn't expect an `if` block to have such limitations. I expect this kind of thing to fail during a validation phase when, for example, a box annotated with pauli-twirling attempts to have overlapping entanglers.

```suggestion
* a group of possibly-nested circuit instructions,
```

Reviewer 3: Yeah, the sentence as I wrote it is a mess.  I moved the bit about nesting out of the list, and modified the top bit to call them "consecutive and/or parallel instructions", though there's probably some more formal graph-theory terminology that I can't bring to mind.  Done in 5e579ba.

* which introduces a new scope for declared variables,
* which can have internal scheduling within itself, but is scheduled atomically within the circuit (i.e. all wires touched within the box are delayed to match the critical path length),
* which forbids optimisations from crossing from outside to inside the box or vice-versa, but permits optimisations across the entire box,
* and can have arbitrary backend- and SDK-agnostic annotations attached to it.

> A `box` can contain any other Qiskit circuit instruction.

Reviewer 2: Redundant
```suggestion
```

> Aside from variable scoping concerns and the resolution of delay lengths, the execution of a circuit containing a `box` should be the same as the execution of the same circuit with the boxed instructions inlined into the circuit.

Reviewer 2: This sentence doesn't look True. for example,

```
circuit.h(0)
with circuit.box():
    circuit.h(0)
    circuit.h(1)
```

has no variable scoping or delays, the only two caveats mentioned, but I might reasonably expect it to be equivalent to 

```
circuit.h(0)
circuit.barrier(0, 1)
circuit.h(0)
circuit.h(1)
```

rather than

```
circuit.h(0)
circuit.h(0)
circuit.h(1)
```

at execution time.

So, does the sentence mean to put "logical" in there somewhere, or should it talk a bit more about barrier sandwiches?

Reviewer 3: Fixed in 5e579ba to describe the walls as synchronisation points with no associated duration.

    def replace_blocks(self, blocks):
        assert len(blocks) == 1
        return Box(blocks[0], self.annotations.copy())
```

See below for more considerations about the ``annotations`` parameter.

This then implies that Qiskit should implement the control-flow builder interfaces for `box`.
Again, using some dummy "annotations" (which are not specified by this RFC, and are given for illustrative purposes only):
```python
from qiskit.circuit import QuantumCircuit

# Hypothetical core-Qiskit box-level annotation.
from qiskit.transpiler.contexts import ScheduleStrategy
# Hypothetical annotations for IBM-specific primitives.
from qiskit_ibm_runtime.contexts import TwirlingStrategy

qc = QuantumCircuit(5, 5)
with qc.box([ScheduleStrategy.ALAP, TwirlingStrategy(...)]):
    qc.h(0)
    qc.cx(0, 1)
    qc.x(2)
with qc.box([ScheduleStrategy.ASAP, TwirlingStrategy(...)]):
    qc.h(2)
    qc.cx(2, 3)
    qc.x(4)
qc.measure(qc.qubits, qc.clbits)
```

For `box` in particular (but also for other control-flow operations), it can be useful to mark certain qubits and other circuit resources as being explicitly "used" by a block, even if the actual operation is a no-op.

> For example, twirling and noise learning may want to consider certain qubits along with a complete block, even if that qubit is undergoing the logical identity for the span.

Reviewer 4: If a box is control-flow block of some width with a non-zero number of idle qubits the transpiler doesn't reduce the width of the circuit does it? It may cause idle qubits to become non-idle when transpiling within the block, however, so is the noop an indication to preserve idles for the duration of the box?

Reviewer: No, it won't reduce the width of the circuit. Passes will leave the box untouched until a pass that understands the annotation encounters it. The noop is a signal that the qubit it acts on can be touched in the box.

This can be worked around by inserting explicit `id` gates into the box, or some other form of no-op operation, but these instructions can sometimes be treated by hardware as something other than a no-op, which might cause performance problems if the compiler fails to optimise them out or understand their intention.
For example, IBM hardware has, in the past, treated the `id` instruction as a single-cycle delay on a single qubit, which is not precisely equal to a no-op for the purposes of scheduling, especially if the rest of the block has a natural duration of zero cycles (such as if it comprises only virtual phase update `rz` gates).
We also do not currently have the concept of a no-op on classical data.
To make the user intent clearer here, we propose adding a `QuantumCircuit.noop` to make the intent explicit:

> > [!NOTE]
> > *Bikeshedding potential*: do you prefer `QuantumCircuit.use` or `QuantumCircuit.noop` for this?
> > `use` is perhaps clearer for more users, but its name only makes sense in a control-flow builder context, whereas `noop` makes sense to appear in the root scope of a `QuantumCircuit`.
> 

Author: We should probably remove this before opening the PR

```suggestion
```

Reviewer 2: i vote to open the rfc with `noop`

```python
from qiskit.circuit import QuantumCircuit
qc = QuantumCircuit(3, 3)
with qc.box():
    # This `box` will be full-width on the circuit, despite every qubit and clbit undergoing the identity.
    qc.noop(*qc.qubits, *qc.clbits)
```

> The `noop` (or `use`) instruction will do the standard Qiskit circuit error checking to validate its inputs are valid resources for the circuit, but will not cause a new entry in the instruction list.

Reviewer 2: ```suggestion
The `noop` instruction will do the standard Qiskit circuit error checking to validate its inputs are valid resources for the circuit, but will not cause a new entry in the instruction list.
```

We do also have a potential option in that core Qiskit can specify a small number of "special case" annotations, which all passes are required to interpret and respect.
This gets tricky from a backwards compatibility perspective, though - if we add a new one of these special annotations in Qiskit 1.3, say, it might invalidate code that worked happily in Qiskit 1.2.

### Serialisation and deserialisation of annotations

If we allow arbitrary annotations, we reasonably need some way of constructing / deconstructing these, without sending arbitrary code to be executed over the wire - that would be a huge security hole.
This is the problem that QPY solves for circuits.
The general principle is that we can rely on reconstruction on the far side of a network boundary, if the receiver only runs code that it chose to install ahead of time.
`pickle` is not suitable for this.

In this model, QPY would need to be extended in a way that allows the serialisation and deserialisation of these attributes.
Two ways of making this extension would be:

* have a plugin interface to QPY, where installed packages can register additional serialisers / deserialisers for annotations.

> * add additional keyword arguments to the `dump`/`load` functions that allow a user to pass in the additional serialisers / deserialisers manually (like how is done for JSON).

Reviewer 2: i'd vote for the kwargs on dump/load for custom serializers.

since the same annotations would be showing up repeatedly in the same circuit, perhaps it could cache by id like pickle.

Reviewer 3: I added a note mentioning your (and my) preference, but explicitly punted the final decision out of this RFC.  We have the main outline of how to do this, and we can flesh out the details fully during implementation on Qiskit.

The transpiler should fail if a user or a pass in its pass manager inserts annotations on boxes that the backend does not support, and are not removed by the end of the transpilation.
We need a way to communicate this information from backends to Qiskit SDK.

For example, an advanced user might attempt to compile an `Estimator` pub with explicit IBM-specific markings of how the noise learning should be performed.
The compilation of a pub represents partial compilation for a particular quantum computer, which will in turn perform further processing to turn this into the set of circuits that will actually be run on the QPU, and so Qiskit SDK's transpiler perhaps ought to have a similar specification of what constitutes "valid" output of such a compilation for quantum-computer processing, much like `Target` represents what constitues valid output for QPU processing.

#### Very very rough sketch of how this might motivate a `BackendV3`

> [!NOTE]
> This is a _very vague_ sketch, I'm just trying to make sure we have a clear definition of what a `Target` is, and we don't start having it become so abstract that it is no longer a good model for a QPU.

> > This is absolutely not necessary to finalise for this RFC, but I feel like it's worth thinking about this as a joined-up story, since previously we have treated the output of `transpile` as "ready for QPU execution", but annotations stymie that.

Reviewer 2: There are other things that stymie that too at the moment, like parameter expressions, which the quantum computer handles but the QPU doesn't. All in all, I really like the direction this section is going in terms of making the distinction between QPU<->Target, Quantum Computer<->Backend nice and crisp.

As for the `sample` and `estimate` methods on backend, I think my biggest concern here is whiplash: we just yoinked `run` away, and this would look like backpedaling. It seems like we could alternately make primitives V3 have to own a backend V3 instance.

Reviewer 3: (For tracking purposes: this is effectively a duplicate thread of https://github.com/SamFerracin/RFCs/pull/1#discussion_r1876733184.)

* the QPY API could be extended to retrieve the serialisers/deserialisers from a `backend` object, rather than needing a user to construct it themselves.

If we _truly_ associated the Qiskit object `Backend` with a "quantum computer" (as opposed to `Target`'s QPU), we might even consider an API where the primitives entry points are moved onto the `BackendV3` object, as in:

```python
class BackendV3:
    @property
    @abstractmethod
    def target(self) -> Target: ...
    @abstractmethod
    def allowed_annotations(self) -> AnnotationChecker: ...
    @abstractmethod
    def sample(self, sampler_pub_likes) -> SamplerJob: ...
    @abstractmethod
    def estimate(self, estimator_pub_likes) -> EstimatorJob: ...
```

This might naturally extend to such a `BackendV3` providing the additional information that is needed for Qiskit SDK to safely transpile entire sampler and estimator pubs; the compilation of a pub _also_ implies further processing will be done by a quantum computer, just like an annotation remaining in a single circuit.

## Case study: twirling, noise learning, and mitigation with `qiskit-ibm-runtime`

> We envision two main workflows that users will follow once `Box`es become available, one for regular users and another one for power users:

Reviewer 2: I don't think this sentence should be talking about `box` users in generality, it should be specifically talking about twirling box users; generic box users' workflow might look totally different.
Instead, this case study needs a short paragraph intro that is twirling/learning/mitigation specific. eg "An application for annotated boxes is to declare to an execution stack that a set of operations should ..blah blah blah"

Reviewer 2: Reading through the whole section, I think we need to:
 - add examples (with disclaimer that we are not attempting to define twirling annotations in this PR, but just give a flavour)
 - better connect to everything written above, especially transpiler passes. what happens with a layout pass if a user *does* happen to create a twirling box manually and then invoke the stock pm?

1. The workflow for regular users:
    - Users initialize a `QuantumCircuit` without boxes, as they do today.
    - They apply all of the desired transpiler passes, for example to map the circuit to an ISA circuit for the backend that they wish to use.
    - They use (a convenience method built around) a new transpiler pass that collects the circuit's gates into boxes, with the ability of specifying different collection strategies.
    - They submit their job.
2. The workflow for power users:
    - Users initialize a `QuantumCircuit` adding boxes manually as they wish.
    - They submit their job.

Setting up these two workflows presents notable advantages, such as:

> - All of the existing transpiler passes can be asked (at least initially) to simply ignore the boxes, since transpilation is meant to happen before the boxes are generated. This can save quite a lot of time that would otherwise be spent adding logic to every single transpiler pass -- and yet, we can choose to do this in the future.

Author: This can probably be removed, now that we use control flow ops?

Reviewer: it might be worthwhile to instead say `to simply ignore the boxes with twirling annotations` or some verbiage to connect it back to [this line](https://github.com/SamFerracin/RFCs/pull/1/files#diff-8468c13b6726725a9dfdf5968d828e7a52b13e30d77470e9e6f50d126179d56aR254)

Reviewer 3: I didn't touch the text in this section - when I'm back, I can review any changes in here that you want to make in the meantime?

> - Users of the primitives can inspect the boxes before submitting their jobs. This is a big improvement: today, the box generation happens in the server and can be unintuitive.

Reviewer: again, maybe reword in terms of annotations

In addition to supporting the two workflow above, we believe that the following workflow should *not* be supported:

3. The workflow for "lazy" users:
    - Users initialize a `QuantumCircuit` without boxes, as they do today.
    - They apply all of the desired transiler passes, for example to map the circuit to an ISA circuit for the backend that they wish to use.
    - They submit their job.

>     - The server creates boxes for the users "behind the scenes."

Reviewer 2: while important, this comment i think has now strayed far from the goal of this RFC since it's about the runtime service implementation and features rather than boxes and annotations. we could instead highlight that boxes bring the possibility of making such operations transparent while still allowing the tight sampling loops to happen near-time

    @abstractmethod
    def sample(self, sampler_pub_likes) -> SamplerJob: ...
    @abstractmethod
    def estimate(self, estimator_pub_likes) -> EstimatorJob: ...
```

This might naturally extend to such a `BackendV3` providing the additional information that is needed for Qiskit SDK to safely transpile entire sampler and estimator pubs; the compilation of a pub _also_ implies further processing will be done by a quantum computer, just like an annotation remaining in a single circuit.

## Case study: twirling, noise learning, and mitigation

Twirling is the process of taking a single base circuit and randomly inserting gates from some distribution, composing them with existing gates, such that the unitary implemented by the circuit and its structure are unchanged. When data is averaged over the result of executing many such independent randomizations, this has the effect of modifiying the noise profile of the circuit into a more tractable one, such as removing the coherent parts of the noise model.

There are many choices for how one might choose to perform these randomizations, where the random gates should be inserted, and with what existing gates they should be composed with.

> If this process is completely automated by the execution stack, the the user needs to be able to clearly specify how these choices should be made.

Reviewer: ```suggestion
If this process is completely automated by the execution stack, then the user needs to be able to clearly specify how these choices should be made.
```

This is especially important in the era of utility because as clients start running larger and larger circuits, it becomes apparent that some twirling choices are better than others.
Moreover, when twirling is used in conjunction with mitigation techniques that rely on noise learning, there needs to be a transparent correspondance between how portions of an application circuit are twirled with how those portions are independently characterized by noise learning techniques.

Therefore, one application of boxes and annotations is to let users declare how they would like to perform twirling, for example with respect to how the gates are randomly drawn, where they are inserted, and what they are composed with.
Once a circuit has been decorated with these boxes and annotations, it is a self-contained description of how twirling should take place.

> > **_NOTE:_**  In what follows, we attempt to give a flavor of what twirling annotations might look like in `qiskit_ibm_runtime`. They will not be defined by the `qiskit` library itself. Take the particular syntax or conventions of the annotations as preliminary.

Reviewer 2: ```suggestion
> **_NOTE:_**  In what follows, we attempt to give a flavor of what twirling annotations might look like in `qiskit_ibm_runtime`. They will not be defined by the `qiskit` library itself. The particular syntax and conventions used here are preliminary and do necessarily represent what will come.
```

```python
from qiskit.circuit import Parameter, QuantumCircuit
from qiskit_ibm_runtime.annotations import PauliTwirl

circuit = QuantumCircuit(6)

# define a twirled layer of CZs
with circuit.box([PauliTwirl(<specifiers>)]):
    # both of these single-qubit gates will be composed with the random 
    # Pauli selections
    circuit.sx(0)
    circuit.rz(Parameter("p"), 0)
    
    circuit.cz(0, 1)
    circuit.cz(2, 3)
    circuit.noop(4, 5)

# define a twirled layer of measurements
with circuit.box([PauliTwirl(<specifiers>)]):
    circuit.measure_all()
```

> We envision two main workflows that users will follow once `box`es and annotations become available to specify twirling, one for regular users and another one for power users:

Reviewer 2: ```suggestion
We envision two main workflows that users will follow once `box`es and annotations become available to specify twirling, one for regular users and another for power users:
```

This `Box` should be able to pass through the Qiskit compiler (under certain limitations), be QPY serialisable, and be able to be transmitted up and down execution stacks.
We will examine how this `Box` might also permit downstream "annotations", outside of the control of the Qiskit SDK compiler, and how the compiler will still be able to reason about such annotations.

The nearest-term use cases for this are:

* have a grouping construction to simplify stretch-based scheduling in Qiskit SDK
* allow advanced users to specify the noise-learning and twirling blocks IBM Primitives error mitigation should use, and the per-block strategy
* allow regular users to see what blocks the IBM Primitives divided their circuit into for noise learning and twirling

## `Box`: a one-off grouping of instructions

As in OpenQASM 3, a `Box` will be:

* a group of non-overlapping (but potentially nested) circuit instructions,
* which introduces a new scope for declared variables,
* which can have internal scheduling within itself, but is scheduled atomically within the circuit (i.e. all wires touched within the box are delayed to match the critical path length),

> * which forbids optimisations from crossing from outside to inside the box or vice-versa, but permits optimisations across the entire box,

Reviewer 5: The regular and power workflows below assume that optimization happens before the assignment of boxes.

Reviewer 2: Although we want the boxes for twirling and mitigation, this RFC defines them more generally, as a core generic construct in qiskit. So the workflows below are specific to the use-case of twirling, but the construct itself needs to have this sort of generic interaction with the transpiler well-defined.

If we go this route, this might motivate a `BackendV3` definition which includes information on what annotations should be permitted to remain.
This has some benefits:

* the transpiler can reject the output from malformed pass-manager pipelines, or input circuits where the user put on a custom annotation that the backend does not know how to resolve.  This allows client-side validation of the circuits.
* the QPY API could be extended to retrieve the serialisers/deserialisers from a `backend` object, rather than needing a user to construct it themselves.

If we _truly_ associated the Qiskit object `Backend` with a "quantum computer" (as opposed to `Target`'s QPU), we might even consider an API where the primitives entry points are moved onto the `BackendV3` object, as in:

```python
class BackendV3:
    @property
    @abstractmethod
    def target(self) -> Target: ...
    @abstractmethod
    def allowed_annotations(self) -> AnnotationChecker: ...
    @abstractmethod
    def sample(self, sampler_pub_likes) -> SamplerJob: ...
    @abstractmethod

>     def estimate(self, estimator_pub_likes) -> EstimatorJob: ...

Reviewer 2: My biggest concern here is perceived whiplash from users who were migrated away from `backend.run` to separate primitive classes, only to be ushered back to backend execution methods. 

I do agree, however, that having the correspondence QuantumComputer feature set <-> Backend, QPU feature set<-> Target is nice and clean.

Reviewer 3: I need to remove this bit that mentions mixing-in of the "execution" methods (`sample`, `estimate`) - it was part of what helped me structure things in my mind on the first pass, but leaving the text there is just going to be a huge distraction from the RFC.  As you say, we can't actually put the execution methods back onto the `Backend` class unless we wanted to alienate every user who just migrated.

Reviewer 3: Done in f2a6e5c.

Setting up these two workflows presents notable advantages, such as:
- All of the existing transpiler passes can be asked (at least initially) to simply ignore the boxes, since transpilation is meant to happen before the boxes are generated. 
- Users of the primitives can inspect the boxes before submitting their jobs. This is a big improvement: today, the box generation happens in the server and can be unintuitive.

In addition to supporting the two workflow above, we believe that the following workflow should *not* be supported:

3. The workflow for "lazy" users:
    - Users initialize a `QuantumCircuit` without boxes, as they do today.
    - They apply all of the desired transiler passes, for example to map the circuit to an ISA circuit for the backend that they wish to use.
    - They submit their job.
    - The server creates boxes for the users "behind the scenes."
  
This workflow essentially takes the same steps as are taken today and therefore presents the same disadvantages discussed above, and also violates the principle "don't do hidden transpilation server-side".

> Another application of boxes is to let users declare how they would like to perform noise learning and mitigation. To do so, they can assign a unique identifier to a box, and subsequently use those identifiers in the primitives' options.

Author: @ihincks FYI

Reviewer 2: we can probably be slightly more specific

> Another application of boxes is to let users declare how they would like to perform noise learning and mitigation. To do so, they can assign a unique identifier to a box via a separate UID annotation, and subsequently provide the primitives' options with a map from those identifiers to noise models they wish to attach to the identifier.

This might naturally extend to such a `BackendV3` providing the additional information that is needed for Qiskit SDK to safely transpile entire sampler and estimator pubs; the compilation of a pub _also_ implies further processing will be done by a quantum computer, just like an annotation remaining in a single circuit.

> ## Case study: twirling, noise learning, and mitigation

Reviewer: Maybe out of scope but could also talk about how PEA can be represented in this way? That is, emitting a Pauli error with rates multiplied by `num_amplifications`?

I remember there was some work @SamFerracin did where only a subset of entangling gates were amplified, this could be a way of specifying said subset (though deterministically as opposed to randomized here).
